### ISSUE TITLE
[PATCH] Support for Basic Authentication in t.w.xmlrpc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11078]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)| @oubiwann reported|
|-|-|
|Trac ID|trac#1474|
|Type|enhancement|
|Created|2006-02-08 13:18:21Z|

Attachments:

* [xmlrpc_basicauth.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fcd/fcd8b5b9ecb89e65d56504f6f6cfe82eed26887e/e2305a89bae75d6e1061a0b65b0c0cda13f0fe68.patch) (3383 bytes) - added by oubiwann on 2006-02-08 13:18:00Z - 
* [test_xmlrpc_basicauth.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fcd/fcd8b5b9ecb89e65d56504f6f6cfe82eed26887e/278a53bf36c34cfb412da6fb2ab76f420f37cc24.patch) (3040 bytes) - added by oubiwann on 2006-03-22 22:20:38Z - New test patch that includes header checks.

<details><summary>Searchable metadata</summary>

```
trac-id__1474 1474
type__enhancement enhancement
reporter__oubiwann oubiwann
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1139404701000000 1139404701000000
changetime__1143325541000000 1143325541000000
version__ 
owner__ 
cc__radix cc__itamarst cc__oubiwann
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

```
#!html
<pre>
This patch adds support for Basic Authentication in twisted.web.xmlrpc by:

1) modifying xmlrpc.Proxy to parse for username and password
2) modifying xmlrpc.QueryFactory to accept named parameters 'user' and 
'password' in the signature in such a way as to be backwards compatible 
(there were no previous named parameters in the sig); and,
3) modifying xmlrpc.QueryProtocol to check if the factory has a 'user' 
attribute, and if so, to set the 'Authentication' header.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

```
#!html
<pre>
One more note: this was created with trunk, revision 15841.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

```
#!html
<pre>
Had an email exachange with Itamar. I will be updating this patch with 
trial tests.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

```
#!html
<pre>
Added a patch for t.w.test.test_xmlrpc. Tests appropiately fail against 
currentl t.w.xmlrpc and pass for the patched t.w.xmlrpc.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Moshe and I looked at this today, and he suggested that we use urllib2.HTTPPasswordMgr. Also, to make a request and only authenticate when a 401 is returned.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

Only sending password on second request is a good idea in theory, but a bad idea so long as we're doing one query per-HTTP connection, because it'll end up being rather slow (especially with SSL).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

The test seems a bit lacking; it doesn't seem to actually test that server receives real HTTP auth headers?

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Yes, very true. I had a hard time figuring out how to do that with the current test setup and they way the classes are written (i.e., no request object to easily get header info). Hmmm. I guess I could just just to a t.w.client.getPage, independent of the actual code in xmlrpc... I didn't think about that.

Also, there's a line missing in the test that would confirm the same tests for unauth are passing for auth. I've added that line. Will update the patch with header tests.

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Crap, that won't work -- it'll circumvent the xmlrpc header-setting. Still looking for a way to test header info... suggestions welcome ;-)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

Wrap the XML-RPC resource with another resource that checks for HTTP auth headers?

**Comment by twisted-trac** [comment]
|[<img alt="oubiwann's avatar" src="https://avatars.githubusercontent.com/u/894251?s=50" width="50" height="50">](https://github.com/oubiwann)|@oubiwann commented|
|-|-|

Itamar, thanks for the heads-up on the wrapper. I wrapped Test and used render() to set self.request, added an xmlrpc_authheader call to the wrapper and then just added an extra check in XMLRPCTestAuthenticated.testAuthHeader.

Tests against the auth header are happening and passing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16446]) Apply Duncan McGreggor's patch to add basic auth support to t.w.xmlrpc.Proxy

Author: Duncan McGreggor
Reviewers: itamarst, exarkun
Fixes [#11078](https://github.com/twisted/twisted/issues/11078)

This adds support for parsing username/password info out of URLs as well as
explicit user/password arguments to Proxy, and then using that information
to do http basic authentication.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
